lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2761,6 +2774,25 @@ ClangModuleUnit *ClangImporter::Implementation::getWrapperForModule(
   if (ClangModuleUnit *cached = cacheEntry.getPointer())
     return cached;
 
+  // Ensure system's C++ stdlib modules aren't imported when
+  // using a custom libc++ on Windows. This is specifically needed
+  // on Windows, as the MSVC C++ module is located in a directory
+  // shared with other system modules, and thus could be still found using
+  // Clang's header search. We use the 'std' module name as we assume
+  // that the custom libc++ doesn't provide its own 'std' module (as such
+  // module would fail to get built due to a module name ambiguity between
+  // MSVC 'std' and libc++'s 'std'). That generally means that the module
+  // map shipped with custom libc++ is expected to be broken into 'std_*'
+  // top level modules, and also have a renamed top-level 'std' module.
```

Comment:
I'm not sure if it is a reasonable expectation that clients would rename `module std` in their libc++ installation. Could we instead rename Swift's `module std` in `GetWindowsFileMappings` by doing, for instance, a find-and-replace in `ucrt.modulemap` and renaming `module std` into something else?

Reply:
Well, I think that we want to preserve the name for compatibility. The default library should be `std`. However, we should think through this - this is a problem for Linux environments where you can select between the different C++ runtimes. We could be on a system with libstdc++ and stlport for example.

Reply:
Sorry I don't follow, why would clients need to access the platform default `std` module when they're building with a different C++ stdlib? On Linux we intentionally made that impossible (modules of both libc++ and libstdc++ are called `std` and they aren't accessible at the same time), to prevent someone from accidentally using two C++ stdlibs at a time and running into ABI issues.

Reply:
If the module name is changed, the `import` statements need to change. The purpose for C++ having the standard is so that any C++ runtime should be a drop in replacement. When we rename the module, this is no longer really the case. We should have an alternative means of selecting the C++ runtime IMO - something akin to `-stdlib=` in `clang++`, perhaps `-c++-stdlib=` being passed to the driver to select between various runtimes. This would also then subsume the need for passing a flag to the linker as the driver is aware of the C++ runtime in use.

Reply:
> If the module name is changed, the import statements need to change. The purpose for C++ having the standard is so that any C++ runtime should be a drop in replacement. When we rename the module, this is no longer really the case.

Ah, that was exactly my point, yes 🙂
I think the module's name should be `std`, not `custom_std`, and the compiler should be smart enough to make sure only one of the modules gets into the search paths.

> something akin to -stdlib= in clang++

This is the model we're currently using in Swift on Linux, yes: `-Xcc -stdlib=libc++`.

Looks like the Clang driver on Windows doesn't understand this flag yet though – re-landing https://github.com/llvm/llvm-project/commit/a4230319f7af786f4d947f39a870f339b32a6fe7 should in theory solve that.

Reply:
Ah, so we are in agreement in concept, just not in words!

> This is the model we're currently using in Swift on Linux, yes: -Xcc -stdlib=libc++.

Right, I'm suggesting that we promote this to a proper driver level flag `-c++-stdlib=`.

> Looks like the Clang driver on Windows doesn't understand this flag yet though – re-landing https://github.com/llvm/llvm-project/commit/a4230319f7af786f4d947f39a870f339b32a6fe7 should in theory solve that.

Why can we not have nice things? Well, at least it is underway. We should be able to cherry-pick that back I suppose.

Reply:
I don't think we need to preserve the compatibility name of `std` that's present in the vcruntime.modulemap when libc++ is used, as you shouldn't be importing that specific module into Swift when using libc++ on windows, so I think renaming the system's `std` automatically instead of asking the user to rename their `std` makes more sense.